File: packages/emitter/testcases/common/functions/delegates/ActionFunc.ts
import { int } from "@tsonic/core/types.js";

// Action types (void return)
export function runAction(action: () => void): void {
  action();
}

export function runActionWithArg(action: (x: int) => void, value: int): void {
  action(value);
}

// Func types (with return)
export function applyFunc<T, R>(fn: (arg: T) => R, value: T): R {
  return fn(value);
}

export function applyFunc2<T1, T2, R>(
  fn: (a: T1, b: T2) => R,
  a: T1,
  b: T2
): R {
  return fn(a, b);
}

// Composing functions
export function compose<A, B, C>(f: (a: A) => B, g: (b: B) => C): (a: A) => C {
  return (a: A): C => g(f(a));
}

File: packages/emitter/testcases/common/expected/functions/delegates/ActionFunc.cs
namespace TestCases.common.functions.delegates
{
        public static class ActionFunc
        {
            public static void runAction(global::System.Action action)
                {
                action();
                }

            public static void runActionWithArg(global::System.Action<int> action, int value)
                {
                action(value);
                }

            public static R applyFunc<T, R>(global::System.Func<T, R> fn, T value)
                {
                return fn(value);
                }

            public static R applyFunc2<T1, T2, R>(global::System.Func<T1, T2, R> fn, T1 a, T2 b)
                {
                return fn(a, b);
                }

            public static global::System.Func<A, C> compose<A, B, C>(global::System.Func<A, B> f, global::System.Func<B, C> g)
                {
                return (a) => g(f(a));
                }
        }
}